File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        res = TreeNode(-1)
        dummy = res
        stack = []
        if root is None:
            return stack
        stack.append(root)
        while stack:
            node = stack.pop()
            res.right = node
            res.left = None
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
            res = res.right
        return dummy.right

        

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
class Solution {
    private TreeNode prev = null;

    private void flattenHelper(TreeNode root) {
        if (root == null) return;

        flattenHelper(root.right);

        flattenHelper(root.left);

        root.right = prev;
        root.left = null;
        prev = root;
    }

    public void flatten(TreeNode root) {
        prev = null;
        flattenHelper(root);
    }
}
